src/test/shared/cloudformation/templateSymbolResolver.test.ts

```diff
@@ -0,0 +1,94 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as assert from 'assert'
+import { instance, mock, when } from 'ts-mockito'
+import * as vscode from 'vscode'
+import {
+    TemplateFunctionResource,
+    TemplateSymbolProvider,
+    TemplateSymbolResolver
+} from '../../../shared/cloudformation/templateSymbolResolver'
+
+const range = new vscode.Range(0, 0, 0, 0)
+
+const firstFunctionType = string('Type')
+const firstFunction = module('Function1', [firstFunctionType])
+
+const secondFunctionType = string('Type')
+const secondFunction = module('Function2', [secondFunctionType])
+
+const looksLikeFunctionType = string('Type')
+const looksLikeFunction = module('LooksLikeFunction', [looksLikeFunctionType])
+
+const symbols: vscode.DocumentSymbol[] = [
+    module('Globals'),
+    module('Resources', [
+        firstFunction,
+        module('NumberType', [number('Type')]),
+        secondFunction,
+        module('MissingType', [string('Missing')]),
+        module('AnotherResource'),
+        looksLikeFunction
+    ])
+]
+
+function module(name: string, children: vscode.DocumentSymbol[] = []) {
+    return symbol(name, vscode.SymbolKind.Module, children)
+}
+
+function string(name: string, children: vscode.DocumentSymbol[] = []) {
+    return symbol(name, vscode.SymbolKind.String, children)
+}
+
+function number(name: string, children: vscode.DocumentSymbol[] = []) {
+    return symbol(name, vscode.SymbolKind.Number, children)
+}
+
+function symbol(name: string, kind: vscode.SymbolKind, children: vscode.DocumentSymbol[] = []) {
+    const newSymbol = new vscode.DocumentSymbol(name, 'detail', kind, range, range)
+    newSymbol.children = children
+
+    return newSymbol
+}
+
+describe('TemplateSymbolResolver', () => {
+    let mockDocument: vscode.TextDocument
+    let document: vscode.TextDocument
+
+    let mockSymbolProvider: TemplateSymbolProvider
+    let symbolProvider: TemplateSymbolProvider
+
+    beforeEach(() => {
+        mockDocument = mock()
+        document = instance(mockDocument)
```

Comment:
I'm not familiar with this pattern. Is there a reason to have the `mockDocument` around?

Reply:
It's the annoying thing about ts-mockito. It makes a distinction between the mock you create and the instance of the mock that you use for injection.

You create expectations and verifications on the former and you inject the latter.

It's an annoying difference compared to normal mockito where there is no distinction. But the library is still very good and the best of the alternatives I looked into. Especially assuming an existing familiarity with mockito by devs.